### DISCUSSION TITLE
App built on mac now leads to "No module named 'encodings' " due to updated package and its accompanying data files.

### DISCUSSION THREAD

**Discussion by fishbacp** [pyinstaller/pyinstaller#6422] [Help]
I'm using pyinstaller 4.7 with python 3.9 on a Mac running 11.6.1. Until quite recently, I had no problems creating an app, `main_app`, that incorporated the mne signal processing package. However, after updating mne, I started to encounter error messages when running pyinstaller, which indicated mne data files were missing. I went into my spec file and added the following at the top,

```
from PyInstaller.utils.hooks import collect_data_files
import os.path

```
After instantiating my Analysis object `a` , where I specify` pathex`, set `datas=[]`, list `hidden imports`, etc., I added to datas a background.png is an image I've always used for my app:

```
data=[('background.png','/Users/fishbacp/Desktop/background.png','DATA')]

```
Because the above is an ordered triple, I added the mne data files as triples also:

```
data_mne=collect_data_files('mne'). # A list of ordered pairs for mne data files

for d in data_mne:
    # d[0], the first entry in the ordered pair d denotes the FULL path to the data file.
    file_name=os.path.basename(d[0]) # Extract the file name from the full path.
    data+= [(file_name, d[0], 'DATA')] # Add, ordered triple of form, (file_name, full path to file_name, 'DATA')

a.datas=data

```
Typing `pyinstaller main_app.spec main_app.py` at the terminal , everything runs fine, modulo the usual warning messages. 

When I try to run the app from the terminal, I encountered the following error message:

```
Fatal Python error: init_fs_encoding: failed to get the Python codec of the filesystem encoding
Python runtime state: core initialized
ModuleNotFoundError: No module named 'encodings'

Current thread 0x0000000109303e00 (most recent call first):
<no Python frame>
```
I tried adding `encodings` as a hidden import. This didn't solve the problem. I then went into the application bundle created here as well as the bundle for an older functioning version of the app, `main_app_old`, which I created prior to updating mne. I discovered that in the functioning app version, the encodings module was found in the following directory:

/Users/fishbacp/Desktop/dist/main_app_old/jedi/third_party/typeshed/stdlib/2

In the new app bundle jedi, and hence, encodings, was nowhere to be found.

**Comment by rokm** [comment]
> a.datas=data

You're overwriting `a.datas` with your own entries instead of extending it (i.e., `a.datas+=data`). Therefore, `base_library.zip` ends up missing in your application, and python interpreter cannot find its `encodings` module.

The real question here is why you are modifying `a.datas` in the first place, when you can simply pass the combined list of two-element tuples
```
[('background.png','/Users/fishbacp/Desktop/background.png'), ]  + collect_data_files('mne')
```
to `Analysis` at instantiation via its `datas` argument.

**Reply by fishbacp** [reply]
Thanks for your response. So I removed `a.datas=data`  from below the instantiation of A. For purposes of readability, I added

`data=[('background.png' , '/Users/fishbacp/Desktop/background.png'), ]  + collect_data_files('mne')`

prior to instantiation, and set `datas=data` within the instantiation. 

However, incorporating these changes into my spec, led to the following error when running pyinstaller:

`Security-Alert: try to store file outside of dist-directory. Aborting. '/Users/fishbacp/Desktop/background.png/background.png'`

I don't understand why it included `background.png/background.png` 

Because of this application bundle isn't produced.

**Reply by rokm** [reply]
Ah, I copy&pasted your final data tuple. It should actually be:
`data=[('/Users/fishbacp/Desktop/background.png', '.'), ] + collect_data_files('mne')`

**Reply by fishbacp** [reply]
That works--the app opens both from the terminal and when I double click the program icon!

What does the extra period within quotes you included, denote? I had seen that mentioned elsewhere but didn't quite understand it.

Thank you so very much!

**Reply by rokm** [reply]
It means that the file should go into top-level application directory. If you do `('/Users/fishbacp/Desktop/background.png', 'background.png')`, then PyInstaller actually collects your image into a sub-directory called `background.png` (i.e., `background.png/background.png`). But if you use dot, it ends up collected as `./background.png`, which is what you want.

**Reply by fishbacp** [reply]
What about loading the background for use in my script? Suppose I wanted to use imshow to plot it. What pathname would I use for loading it, i.e. what would ??? be below?

```
    import mathplotlib.pyplot as plt
    fig, ax = plt.subplots()
    background_to_plot= plt.imread(????)
    ax.imshow(background_to_plot)
```
It wouldn't be as simple as` background.png` would it?

**Reply by rokm** [reply]
> It wouldn't be as simple as `background.png` would it?

No, because this works only if the image is located in your current working directory. Which might be where the executable is (and then image will be found) or may not be, depending on `onedir` vs `onefile` mode and how the executable was launched. So you should always use absolute paths and anchor them to either location of your python script (in unfrozen version) or to the frozen application's top-level directory (in frozen version).

In a frozen application, you can use `sys._MEIPASS` to construct paths relative to frozen application's directory, regardless of `onedir` or `onefile` mode (or `os.path.dirname(__file__)` , which will give you the same path, although only in the entry-point script). Read [here](https://pyinstaller.readthedocs.io/en/stable/runtime-information.html#using-file) for more details.

**Reply by rokm** [reply]
So for example:

```python
import sys
import mathplotlib.pyplot as plt

if getattr(sys, 'frozen', False):
    # Frozen application
    img_path = os.path.join(sys._MEIPASS, 'background.png')
else:
    # Original script
    img_path = '/Users/fishbacp/Desktop/background.png'

fig, ax = plt.subplots()
background_to_plot= plt.imread(img_path)
ax.imshow(background_to_plot)
```

**Reply by fishbacp** [reply]
This is what I had done before the mne package update, although I'm not sure as to why it worked. In my main script, I included this:

```
def resource_path(relative_path):
    try:
        base_path = sys._MEIPASS
    except Exception:
        base_path = os.path.abspath(".")

    return os.path.join(base_path, relative_path)

```
and then loaded the image using the above function as follows:

`background_to_plot=plt.imread(resource_path('/Users/fishbacp/Desktop/background.png'))`

**Reply by fishbacp** [reply]
I'm so sorry--our messages crossed. I'm sensing  your suggested code and what I just posted have similar goals?

**Reply by rokm** [reply]
> `background_to_plot=plt.imread(resource_path('/Users/fishbacp/Desktop/background.png'))`

This always read the image from your desktop, `'/Users/fishbacp/Desktop/background.png'`.

**Reply by fishbacp** [reply]
Okay, I see. Just to confirm, in your code above it's written `getattr(sys, 'frozen', False)` when testing for a frozen application? (Or should it be  `getattr(sys, 'frozen', True) ` ?

**Reply by rokm** [reply]
> Okay, I see. Just to confirm, in your code above it's written `getattr(sys, 'frozen', False)` when testing for a frozen application? (Or should it be `getattr(sys, 'frozen', True) ` ?

C'mon, we're not running a python coding school here... read the [docs](https://docs.python.org/3/library/functions.html#getattr) and decide for yourself if you need `True` or `False` there...

**Reply by fishbacp** [reply]
I will. Thanks so much for all your assistance today. It was very much appreciated! 👍
